File: JAVA/PROYECTO_JAVA_PDOO/src/proyecto_java_pdoo/Practica_1.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package proyecto_java_pdoo;

/**
 *
 * @author alejandro
 */
public class Practica_1 {
    
    public enum Directions{
        LEFT,
        RIGHT,
        UP,
        DOWN
    }
    
    public enum Orientation{
        VERTICAL,
        HORIZONTAL
    }
    
    public enum GameCharacter{
        PLAYER,
        MONSTER
    }
    
    
    
    public class Weapon{
        private float power;
        private int uses;
        
        public Weapon(float power, int uses){
            this.power=power;
            this.uses=uses;
        }
        
        public float attack(){
            if(uses>0){
                uses--;
                return power;
            }else{
                return 0;
            }
        }
        
        public String toString(){
            return "W[" + power + "," + uses + "]";
        }
        
        public float getPower(){
            return power;
        }
        
        public int getUses(){
            return uses;
        }
        
        public static boolean discard(){
            return discardElement(usesLeft);
        }
    }
    
    
    public class Shield{
        private float protection;
        private int uses;
        
        public Shield(float protection, int uses){
            this.protection=protection;
            this.uses=uses;
        }
        
        public float protect(){
            if(uses>0){
                uses--;
                return protection;
            }else{
                return 0;
            }
        }
        
        public String toString(){
            return "S[" + protection + "," + uses + "]";
        }
        
        public float getProtection(){
            return protection;
        }
        
        public int getUses(){
            return uses;
        }
        
        public static boolean discard(){
            return discardElement(usesLeft);
        }
    }
    
    public class Dice{
        private static final int MAX_USES=5;
        private static final float MAX_INTELLIGENCE=10.0f;
        private static final float MAX_STRENGTH=10.0f;
        private static final float RESURRECT_PROB=0.3f;
        private static final int WEAPONS_REWARD=2;
        private static final int SHIELDS_REWARD=3;
        private static final int HEALTH_REWARD=5;
        private static final int MAX_ATTACK=3;
        private static final int MAX_SHIELD=2;
        
        
        private static Random generator = new Random();
        
        public static int randomPos(int max){
            return generator.nextInt(max);
        }
        
        public static int whoStarts(int nplayers){
            return generator.nextInt(nplayers);
        }
        
        public static float randomIntelligence(){
            return generator.nextFloat()*MAX_INTELLIGENCE;
        }
        
        public static float randomStrength(){
            return generator.nextFloat()*MAX_STRENGTH;
        }
        
        public static boolean resurrectPlayer(){
            return generator.nextFloat() < RESURRECT_PROB;
        }
        
        public static int weaponsReward(){
            return generator.nextInt(WEAPONS_REWARD + 1);
        }
        
        public static int shieldsReward(){
            return generator.nextInt(SHIELDS_REWARD + 1);
        }
        
        public static int healthReward(){
            return generator.nextInt(HEALTH_REWARD + 1);
        }
        
        public static float weaponPower(){
            return generator.nextFloat()*MAX_ATTACK;
        }
        
        public static float shieldPower(){
            return generator.nextFloat()*MAX_SHIELD;
        }
        
        public static int usesLeft(){
            return generator.nextInt(MAX_USES + 1);
        }
        
        public static float intensity(float competence){
            return generator.nextFloat()*competence;
        }
        
        public static boolean discardElement(int usesLeft){
            if(usesLeft >= MAX_USES){
                return false;
            }else if(usesLeft == 0){
                return true;
            }else{
                float probability = (float)*(MAX_USES - usesLeft) / MAX_USES;
                return generator.nextFloat() < probability;
            }
        }
        
    }
    
    public class GameState{
        private String labyrinth;
        private String players;
        private String monsters;
        private int currentPlayer;
        private boolean winner;
        private String log;
        
        
        public GameState(String labyrinth, String players, String monsters, int currentPlayer, boolean winner, String log){
            this.labyrinth=labyrinth;
            this.players=players;
            this.monsters=monsters;
            this.currentPlayer=currentPlayer;
            this.winner=winner;
            this.log=log;
        }
        
        public String getLabyrinth(){
            return labyrinth;
        }
        
        public String getPlayers(){
            return players;
        }
        
        public String getMonsters(){
            return monsters;
        }
        
        public int getCurrentPlayer(){
            return currentPlayer;
        }
        
        public boolean getWinner(){
            return winner;
        }
        
        public String getLog(){
            return log;
        }
        
        
    }
    
    
}

File: RUBY/PRACTICA1/Practica-1/practica1.rb
module Directions
  LEFT =:left
  RIGHT =:right
  UP =:up
  DOWN =:down
end

module Orientation
  VERTICAL =:vertical
  HORIZONTAL =:horizontal
end

module GameCharacter
  PLAYER =:player
  MONSTER =:monster
end


class Weapon
  attr_reader :power, :uses

  def initialize(power,uses)
    @power = power.to_f
    @uses = uses.to_i
  end

  def attack
    if @uses > 0
      @uses -= 1
      @power
    else
      0
    end
  end

  def to_s
    "W[#{@power},#{@uses}]"
  end

  def discard
    Dice.discard_element(@uses)
  end
end

class Shield
  attr_reader :protection, :uses

  def initialize(protection,uses)
    @protection = protection.to_f
    @uses = uses.to_i
  end

  def protect
    if @uses > 0
      @uses -= 1
      @protection
    else
      0
    end
  end

  def to_s
    "W[#{@protection},#{@uses}]"
  end

  def discard
    Dice.discard_element(@uses)
  end
end

class Dice
  @@MAX_USES = 5
  @@MAX_INTELLIGENCE = 10.0
  @@MAX_STRENGTH = 10.0
  @@RESURRECT_PROB = 0.3
  @@MAX_RESURRECT_PROB = 1.0
  @@WEAPONS_REWARD = 2
  @@SHIELDS_REWARD = 3
  @@HEALTH_REWARD = 5
  @@MAX_ATTACK = 3
  @@MAX_SHIELD = 2

  @@generator = Random.new

  class << self
    attr_reader :MAX_USES, :MAX_INTELLIGENCE, :MAX_STRENGTH
    attr_reader :RESURRECT_PROB, :WEAPONS_REWARD, :SHIELDS_REWARD, :HEALTH_REWARD
    attr_reader :MAX_ATTACK, :MAX_SHIELD

    def roll(sides=6)
      @@generator.rand(1..sides)
    end
  end
end


def self.random_pos(max)
  raise ArgumentError, "EL valor tiene que ser mayor a 0" if max <= 0

  @@generator.rand(0...max)
end

def self.who_starts(nplayers)
  raise ArgumentError, "El numero de jugadores tiene que ser mayor a 0" if nplayers <= 0

  @@generator.rand(0...nplayers)
end

def self.random_intelligente
  @@generator.rand(0...MAX_INTELLIGENCE)
end

def self.random_strength
  @@generator.rand(0...MAX_STRENGTH)
end

def self.resurrect_player
  @@generator.rand(0...MAX_RESURRECT_PROB) < RESURRECT_PROB
end

def self.weapons_reward
  @@generator.rand(0...WEAPONS_REWARD)
end

def self.shields_reward
  @@generator.rand(0...SHIELDS_REWARD)
end

def self.health_reward
  @@generator.rand(0...HEALTH_REWARD)
end

def self.weapon_power
  @@generator.rand(0...MAX_ATTACK)
end

def self.shield_power
  @@generator.rand(0..MAX_SHIELD)
end

def self.uses_left
  @@generator.rand(0..MAX_USES)
end

def self.intensity(competence)
  @@generator.rand(0...competence)
end

def self.discard_element(uses_left)
  if @@generator.rand(1..@@MAX_USES) > uses_left
    return true
  end
  false
end

class GameState
  attr_reader :labyrinth, :players, :monsters, :current_player, :winner, :log

  def initialize(labyrinth, players, monsters, current_player, winner, log)
    @labyrinth = labyrinth
    @players = players
    @monsters = monsters
    @current_player = current_player
    @winner = winner
    @log = log
  end
end